cpp/include/rapidsmpf/topology_discovery.hpp

```diff
@@ -0,0 +1,122 @@
+    std::vector<std::string>
+        network_devices;  ///< Network devices (NICs) optimal for this GPU.
+};
+
+/**
+ * @brief Network device information.
+ */
+struct NetworkDeviceInfo {
+    std::string name;  ///< Device name (e.g., "mlx5_0").
+    int numa_node;  ///< NUMA node ID (-1 if unknown).
+    std::string pci_bus_id;  ///< PCI bus ID.
+};
+
+/**
+ * @brief System topology information.
+ */
+struct SystemTopologyInfo {
+    std::string hostname;  ///< System hostname.
+    unsigned int num_gpus;  ///< Total number of GPUs.
+    int num_numa_nodes;  ///< Total number of NUMA nodes.
+    int num_network_devices;  ///< Total number of network devices.
+    std::vector<GpuTopologyInfo> gpus;  ///< GPU topology information.
+    std::vector<NetworkDeviceInfo> network_devices;  ///< Network device information.
+};
+
+/**
+ * @brief PCIe topology path types.
+ */
+enum class PciePathType {
+    PIX = 0,  ///< Connection traversing at most a single PCIe bridge (best).
+    PXB = 1,  ///< Connection traversing multiple PCIe bridges.
+    PHB = 2,  ///< Connection traversing PCIe Host Bridge.
+    NODE = 3,  ///< Connection traversing PCIe and interconnect within NUMA node.
+    SYS = 4  ///< Connection traversing NUMA interconnect (worst).
+};
+
+/**
+ * @brief Discover system topology including GPUs, NUMA nodes, and network devices.
+ *
+ * This class provides methods to discover system topology information using NVML
+ * and /sys filesystem queries. It dynamically identifies GPU-to-NUMA-to-NIC mappings
+ * based on PCIe topology.
+ *
+ * Example usage:
+ * @code
+ * rapidsmpf::TopologyDiscovery discovery;
+ * if (discovery.discover()) {
+ *     auto topology = discovery.get_topology();
+ * }
+ * @endcode
+ */
+class TopologyDiscovery {
+  public:
+    /**
+     * @brief Construct a TopologyDiscovery instance.
+     */
+    TopologyDiscovery() = default;
+
+    /**
+     * @brief Destroy the TopologyDiscovery instance.
+     */
+    ~TopologyDiscovery() = default;
+
+    /**
+     * @brief Discover system topology.
+     *
+     * This method performs the actual discovery of GPUs, NUMA nodes, CPU affinity,
+     * and network devices. It must be called before `get_topology()`.
+     *
+     * @return true if discovery was successful, false otherwise.
+     */
+    bool discover();
+
+    /**
+     * @brief Get the discovered topology information.
+     *
+     * @return SystemTopologyInfo structure containing all topology data.
+     * @note `discover()` must be called first.
+     */
+    SystemTopologyInfo const& get_topology() const {
+        return topology_;
+    }
+
+    /**
+     * @brief Check if topology has been discovered.
+     *
+     * @return true if `discover()` has been called successfully.
+     */
+    bool is_discovered() const {
+        return discovered_;
+    }
+
+  private:
+    SystemTopologyInfo topology_;  ///< Discovered topology information.
```

Comment:
What about using `std::optional<SystemTopologyInfo>` and removing `discovered_`? 
Or maybe even better, drop this class and have a function that returns a `SystemTopologyInfo` instead?